rotkehlchen/chain/ethereum/modules/thegraph/balances.py

```diff
@@ -0,0 +1,118 @@
+    from rotkehlchen.types import CHAIN_IDS_WITH_BALANCE_PROTOCOLS
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class DelegationData(NamedTuple):
+    shares: FVal  # Shares owned by a delegator in the pool
+    tokensLocked: FVal  # Tokens locked for undelegation
+    tokensLockedUntil: FVal  # Block when locked tokens can be withdrawn
+
+class DelegationPoolData(NamedTuple):
+    cooldownBlocks: FVal  # Blocks to wait before updating parameters
+    indexingRewardCut: FVal  # in PPM
+    queryFeeCut: FVal  # in PPM
+    updatedAtBlock: FVal  # Block when the pool was last updated
+    tokens: FVal  # Total tokens as pool reserves
+    shares: FVal  # Total shares minted in the pool
+
+class ThegrapBalances(ProtocolWithBalance):
+    def __init__(
+            self,
+            database: DBHandler,
+            evm_inquirer: 'EvmNodeInquirer',
+            chain_id: 'CHAIN_IDS_WITH_BALANCE_PROTOCOLS',
+    ):
+        super().__init__(
+            database=database,
+            evm_inquirer=evm_inquirer,
+            chain_id=chain_id,
+            counterparty=CPT_THEGRAPH,
+        )
+        self.token = A_GRT.resolve_to_evm_token()
+
+    def get_deposit_events(self) -> set[tuple[HistoryEventType, HistoryEventSubType]]:
+        return {(HistoryEventType.STAKING, HistoryEventSubType.DEPOSIT_ASSET)}
+
+    def query_balances(self) -> 'BalancesType':
+        balances: BalancesType = defaultdict(lambda: defaultdict(Balance))
+
+        # fetch deposit events
+        addresses_with_deposits = self.addresses_with_deposits(
+            product=None,
+            deposit_events=self.get_deposit_events(),
+        )
+        # remap all events into dictionary: "delegator address" => "list of indexer addresses"
+        indexers_with_shares = defaultdict(list)
+        for address, eventList in addresses_with_deposits.items():
+            for event in eventList:
+                if event.extra_data['indexer'] not in indexers_with_shares[address]:
+                    indexers_with_shares[address].append(event.extra_data['indexer'])
+
+        # query delegation and pools data for every delegator address
+        staking_proxy_contract = self.evm_inquirer.contracts.contract(string_to_evm_address('0xF55041E37E12cD407ad00CE2910B8269B01263b9'))  # noqa: E501
+        # TODO: see if there's a way to do multicalls or query by chunks
+        chunk_size, call_order = get_chunk_size_call_order(self.evm_inquirer)
+        for delegator, indexers in indexers_with_shares:
+            for indexer in indexers:
+                # query how many shares delegator currently have at the indexer/pool
+                stake_result = staking_proxy_contract.call(
+                    node_inquirer=self.evm_inquirer,
+                    method_name='getDelegation',
+                    arguments=[indexer, delegator],
+                    call_order=call_order,
+                )
```

Comment:
you can use multicall here. Check the code for multicall. What it does is execution multiple evm calls in a single rpc contract call. You have to later process the result as a list but is much faster